### PR TITLE
[c10d] Make broadcast as a custom op

### PR DESCRIPTION
**Pull request by alanwaketan**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #75924

Summary:
This is a WIP patch to make broadcast as a custom op such that it's dispatcher
passable.

Test Plan:
python test/distributed/test_c10d_nccl.py -k test_broadcast_ops

### CODE REVIEW

File: torch/csrc/distributed/c10d/ProcessGroup.cpp [hunk 1]
```diff
@@ -186,4 +186,11 @@ void ProcessGroup::init() {
   C10_LOG_API_USAGE_ONCE(fmt::format("c10d.process_group_{}", getBackendName()));
 }
 
+TORCH_LIBRARY(c10d, m) {
+  m.class_<ProcessGroup>("ProcessGroup")
```
Comment (dzhulgakov) [inline torch/csrc/distributed/c10d/ProcessGroup.cpp:190]: If you're still getting import errors you might need to move it into the same file as the operator registration. You require this type registration to run before the operator can be registered. But there's no guaranteed order of execution for C++ static initialization from different modules.

Comment (alanwaketan) [reply torch/csrc/distributed/c10d/ProcessGroup.cpp:190]: Good catch.

Comment (alanwaketan) [reply torch/csrc/distributed/c10d/ProcessGroup.cpp:190]: It solved the issue. Thanks, Dima!

File: aot.py [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+    print(fx_module.code)
+    return fx_module
+
+def demo_basic(rank, world_size):
+    # without flushing, mp won't print this message if crash.
+    print(f"Running basic DDP example on rank {rank}, and process id: {os.getpid()}", flush=True)
+    setup(rank, world_size)
+
+    # Pass on the compiler_fn to the aot_function API
+    aot_print_fn = aot_function(broadcast, fw_compiler=compiler_fn, bw_compiler=compiler_fn)
```
Comment (alanwaketan) [inline aot.py:41]: @Chillee I added example of AOT here. And this is the output. Looks like it works just don't know how to deal with the `bw_compiler`:
```
(pytorch37) jwtan@dev-st-p4d24xlarge-2:/fsx/users/jwtan/work/pytorch$ python aot.py 
main process id: 96435
Running basic DDP example on rank 0, and process id: 96478
Running basic DDP example on rank 1, and process id: 96479
96478 broadcast: tensor([[1., 1., 1.],
        [1., 1., 1.]], device='cuda:0')
96479 broadcast: tensor([[1., 1., 1.],
        [1., 1., 1.]], device='cuda:1')
96478 broadcast: tensor([[1., 1., 1.],
        [1., 1., 1.]], device='cuda:0')
96479 broadcast: tensor([[1., 1., 1.],
        [1., 1., 1.]], device='cuda:1')



def forward(self):
    _tensor_constant0 = self._tensor_constant0
    return [_tensor_constant0]
    



def forward(self, tangents_1):
    return []
    



def forward(self):
    _tensor_constant0 = self._tensor_constant0
    return [_tensor_constant0]
    



def forward(self, tangents_1):
    return []
    
96479 broadcast: tensor([[1., 1., 1.],
        [1., 1., 1.]], device='cuda:1')
96478 broadcast: tensor([[1., 1., 1.],
        [1., 1., 1.]], device='cuda:0')
```

Comment (alanwaketan) [reply aot.py:36]: Looks like AOT doesn't work... Debugging...

Comment (alanwaketan) [reply aot.py:36]: Added a print after __torch_dispatch__ in AOT: https://github.com/pytorch/functorch/blob/main/functorch/_src/python_key.py#L73

Nothing got printed out... So it seems __torch_dispatch__ cannot recognize torch.ops...

**Comment by alanwaketan** [comment]
#76722 is the formal version of this pull request. Let's close this one.  Thanks Edward for helping me out on the dispatch key stuff.

**Pull request closed without merging**